### PR TITLE
Match state machine and robot placement

### PR DESCRIPTION
**Pull request by LucaHaverty**
### Description
A state machine to control match mode. Contains states for things like configuring the match and auto. Each state has a start, update, and end function that are called automatically by MatchStateMachine.cs. The state machine will allow the user to configure, enter, play, and exit a match.

In addition to the state machine, there is a new robot placement system that is used when positioning robots before a match. This allows the user to select a location with their mouse and the robot will automatically be positioned slightly above the field.

### Objectives (mostly different states to create basic functionality for)
- [x] State machine framework
- [x] Add match mode to main menu instead of "coming soon"
- [x] Match config menu (robot and field selection, other settings)
- [x] Spawn all robots without gizmos
- [x] Robot placement state with new placement functionality
- [x] Auto and teleop state
- [x] Post-match results state

### Testing Done
- Robot placement on different fields with different robots
- Traversed through game states to verify that the state machine is functioning as expected
- Tested state machine with multiple robots spawned
- Tested receiving an event from a state starting

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1493)

### CODE REVIEW

**Review by PepperLola — CHANGES_REQUESTED** [review]
I think maybe the `MatchResultsModal` accept button should close the panel and set the match state to `None`? Also rename `MatachResults` to `MatchResults`.

File: engine/Assets/Scripts/Gizmos/GizmoManager.cs [hunk 1]
```diff
@@ -94,7 +94,7 @@ public static void SpawnGizmo(Vector3 position, Quaternion rotation, Action<Tran
             if (_currentGizmoConfig.HasValue)
                 ExitGizmo();
 
-            // Debug.Log("spawn gizmo");
+            Debug.Log("spawn gizmo");
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/Gizmos/GizmoManager.cs:97]: `Debug.Log` still needed?

File: engine/Assets/Scripts/Modes/MatchMode.cs [hunk 2]
```diff
@@ -1,24 +1,42 @@
...
+    private MatchStateMachine _stateMachine;
+    public void Start()
+    {
+        Array.Fill(SelectedRobots, -1);
+        Array.Fill(RobotSpawnLocations, (Vector3.zero, Quaternion.identity));
+        
+        /*SynthesisAPI.EventBus.EventBus.NewTypeListener<MatchStateMachine.OnStateStarted>(e =>
+        {
+            Debug.Log(((MatchStateMachine.OnStateStarted)e).state.GetType());
+        });*/
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/Modes/MatchMode.cs:31]: Can this code block be removed?

File: engine/Assets/Scripts/UI/Dynamic/Modals/Spawning/MatchModeModal.cs [hunk 3]
```diff
@@ -63,43 +68,45 @@ public override void Create()
...
+        /*MainContent.CreateLabel().ApplyTemplate(VerticalLayout).SetText("Select Alliance Color");
         var allianceSelection = MainContent.CreateDropdown().ApplyTemplate(Dropdown.VerticalLayoutTemplate)
             .SetOptions(new string[] { "Red", "Blue" })
-            .AddOnValueChangedEvent((d, i, data) => _allianceColor = i).ApplyTemplate(VerticalLayout);
-
-        /*MainContent.CreateLabel().ApplyTemplate(VerticalLayout).SetText("Select Spawn Position");
-        var spawnPosition = MainContent.CreateDropdown().ApplyTemplate(Dropdown.VerticalLayoutTemplate)
-            .SetOptions(new string[] { "Left", "Middle", "Right" })
-            .AddOnValueChangedEvent((d, i, data) => _spawnPosition = i).ApplyTemplate(VerticalLayout);*/
+            .AddOnValueChangedEvent((d, i, data) => _allianceColor = i).ApplyTemplate(VerticalLayout);*/
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/UI/Dynamic/Modals/Spawning/MatchModeModal.cs:95]: Can this code block be removed?

File: engine/Assets/Scripts/UI/Dynamic/Modals/Spawning/MatchModeModal.cs [hunk 4]
```diff
@@ -113,22 +120,22 @@ public IEnumerator LoadMatch()
...
-                RobotSimObject.SpawnRobot(_robotFiles[_robotIndex]);
+                RobotSimObject.SpawnRobot(_robotFiles[_robotIndex], false);
             }
             
             
-            MatchMode.currentRobotIndex = _robotIndex;
+            MatchMode.CurrentRobotIndex = _robotIndex;
         }
-        Scoring.ResetScore();
+        Scoring.ResetScore();*/
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/UI/Dynamic/Modals/Spawning/MatchModeModal.cs:133]: Can this code block be removed?

File: engine/Assets/Scripts/UI/Dynamic/Panels/Configuring/StartMatchModePanel.cs [hunk 5]
```diff
@@ -9,8 +9,9 @@ public class StartMatchModePanel : PanelDynamic {
         public StartMatchModePanel() : base(new Vector2(300f, 200f)) { }
 
         public override bool Create() {
-            Title.SetText("Almost Ready").SetFontSize(25f);
-            PanelImage.RootGameObject.SetActive(false);
+
+            Title.SetText("Start Match?").SetFontSize(25f);
+            //PanelImage.RootGameObject.SetActive(false);
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/UI/Dynamic/Panels/Configuring/StartMatchModePanel.cs:14]: Can this line be removed?

File: engine/Assets/Scripts/UI/Dynamic/Panels/Configuring/StartMatchModePanel.cs [hunk 6]
```diff
@@ -22,8 +23,8 @@ public class StartMatchModePanel : PanelDynamic {
                 .AddOnClickedEvent(b => {
                     DynamicUIManager.CreateModal<MatchModeModal>();
                 });
-
-            GizmoManager.SpawnGizmo<RobotSimObject>(RobotSimObject.GetCurrentlyPossessedRobot());
+ 
+            //GizmoManager.SpawnGizmo<RobotSimObject>(RobotSimObject.GetCurrentlyPossessedRobot());
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/UI/Dynamic/Panels/Configuring/StartMatchModePanel.cs:27]: Can this line be removed?

File: engine/Assets/Scripts/UI/Dynamic/Panels/SpawnLocationPanel.cs [hunk 7]
```diff
@@ -1,162 +1,259 @@
...
+           // Shooting.ConfigureGamepieces();
+           
+           //TEMPORARY: FOR POWERUP ONLY
+           
+           Scoring.CreatePowerupScoreZones();
+           DynamicUIManager.CloseAllPanels(true);
+           DynamicUIManager.CreatePanel<Synthesis.UI.Dynamic.ScoreboardPanel>(true);
+
+           GizmoManager.ExitGizmo();
+       }*/
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/UI/Dynamic/Panels/SpawnLocationPanel.cs:100]: Can this code block be removed?

File: engine/Assets/Scripts/Modes/MatchStateMachine.cs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        public override void Start() {
+            base.Start();
+            
+            // TODO: start teleop timer on scoreboard
+        }
+
+        public override void Update() {
+            // TEMP END CONDITION FOR STATE MACHINE TESTING
+            if (Input.GetKeyDown(KeyCode.RightArrow))
+                MatchStateMachine.Instance.SetState(StateName.MatchResults);
```
Comment (BrandonPacewic) [inline engine/Assets/Scripts/Modes/MatchStateMachine.cs:189]: Still needed?

**Review by BrandonPacewic — CHANGES_REQUESTED** [review]
Mainly just related to cleanups.

File: engine/Assets/Scripts/Modes/MatchStateMachine.cs (hunk 8 above)
Comment (BrandonPacewic) [reply engine/Assets/Scripts/Modes/MatchStateMachine.cs:189]: 👍

**Review by PepperLola — CHANGES_REQUESTED** [review]
~~I'm not sure if it's an issue on my end, but I wasn't able to rotate the robot without snapping by holding shift. With snapping, it would rotate normally, but when holding shift it wouldn't rotate at all.~~

Nevermind, maybe change snapping from a toggle to a hold control?

**Comment by BrandonPacewic** [comment]
> Nevermind, maybe change snapping from a toggle to a hold control?

Is it a toggle right now? I agree it should be a hold.

**Review by Markytt — APPROVED** [review]

**Review by PepperLola — APPROVED** [review]
Everything looks good now!

**Pull request merged**